src/editor/grid.rs

```diff
@@ -0,0 +1,76 @@
+use std::sync::Arc;
+use log::trace;
+
+use super::style::Style;
+
+type GridCell = Option<(String, Option<Arc<Style>>)>;
+
+pub struct CharacterGrid {
+    pub characters: Vec<GridCell>,
+    pub width: u64,
+    pub height: u64,
+    pub should_clear: bool,
+
+    dirty: Vec<bool>,
+}
+
+impl CharacterGrid {
+    pub fn new() -> CharacterGrid {
+        CharacterGrid {
+            characters: vec![],
+            dirty: vec![],
+            width: 0,
+            height: 0,
+            should_clear: true,
+        }
+    }
+
+    pub fn resize(&mut self, new_size: (u64, u64)) {
+        trace!("Editor resized");
+        self.width = new_size.0;
+        self.height = new_size.1;
+        self.clear();
+    }
+
+    pub fn clear(&mut self) {
+        trace!("Editor cleared");
+        let cell_count = (self.width * self.height) as usize;
+        self.characters = vec![None; cell_count];
+        self.dirty = vec![true; cell_count];
```

Comment:
Couldn't we recycle the previous allocations? We could use [`Vec::resize_with`] to do that.

```suggestion
	    self.characters.clear();
        self.characters.resize_default(cell_count);
        self.dirty.clear();
        self.dirty.resize_with(cell_count, || true); // default of bool is false.
```

[`Vec::resize_with`]: https://doc.rust-lang.org/std/vec/struct.Vec.html#method.resize_with

Reply:
According to the docs, `resize_with` only applies the new values to the newly allocated elements. The old ones retain their old value. Same is true for the other resize functions, actually, which makes my other uses of them wrong

Reply:
> The old ones retain their old value.

Exactly, just call `Vec::clear` before.

Reply:
Huh. Ok, sounds good :+1: